### PR TITLE
Add Kokkos-kernels recipe.

### PR DESCRIPTION
**Pull request by vincentmr**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [ ] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by vincentmr** [comment]
@conda-forge/help-c-cpp, ready for review!

File: .azure-pipelines/azure-pipelines-linux.yml [hunk 1]
```diff
@@ -30,115 +30,3 @@ jobs:
 
   - publish: build_artifacts/noarch/
     artifact: conda_pkgs_noarch
-
```
Comment (dopplershift) [inline .azure-pipelines/azure-pipelines-linux.yml:33]: You should not be making changes to this file.

Comment (vincentmr) [reply .azure-pipelines/azure-pipelines-linux.yml:33]: Hi @dopplershift , if I don't want to build any CUDA backend, should I revert the changes to this file and set some condition in my recipe?

Comment (dopplershift) [reply .azure-pipelines/azure-pipelines-linux.yml]: I'm not an expert on the compiled side here, but IIRC the CUDA builds are opt-in. The CI jobs that ran look like the standard ones to me. @conda-forge/help-c-cpp anything to add?

Comment (vincentmr) [reply .azure-pipelines/azure-pipelines-linux.yml]: I had the impression it was opt-out, but I think you're right. I've just reverted the changes.

**Comment by BastianZim** [comment]
That looks good to me, thank you!

@leofang Would you mind having a look as well? It's a dependency for Pennylane.

File: recipes/kokkos-kernels/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+test:
+  commands:
+  # Lots more CMake files; just checking for one
+    - test -f $PREFIX/lib/cmake/KokkosKernels/KokkosKernelsConfig.cmake  # [unix]
+  # Testing explicitly for all shared libs, but not links
+    - test -f $PREFIX/lib/libkokkoskernels.so  # [linux]
+    - test -f $PREFIX/lib/libkokkoskernels.dylib  # [osx]
+  # There are lots more headers; only test for one.
+    - test -f $PREFIX/include/KokkosKernels_Handle.hpp  # [unix]
+  # There are no binary
```
Comment (leofang) [inline recipes/kokkos-kernels/meta.yaml:39]: It seems to me one can compile some dummy code against `kokkos-kernels`. How hard is it to come up with one? If we can do a simple compilation test here, it'd be great. You'd be surprised how many issues a naive test like this could catch 🙂

**Review by leofang — COMMENTED** [review]
Sorry I missed the ping. LGTM overall. Left a few perhaps naive questions, since I don't know Kokkos well enough.

File: recipes/kokkos-kernels/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+source:
+  url: https://github.com/kokkos/kokkos-kernels/archive/refs/tags/{{ version }}.tar.gz
+  sha256: b2060f5894bdaf7f7d4793b90444fac260460cfa80595afcbcb955518864b446
+
+build:
+  skip: true  # [win]
+  number: 0
+  run_exports:
+    - {{ pin_subpackage('kokkos-kernels') }}
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:16]: I'm not 100% sure, but I would say yes.

File: recipes/kokkos-kernels/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+requirements:
+  build:
+    - cmake >=3.21.1,<4
+    - {{ compiler('cxx') }}
+    - llvm-openmp  # [osx]
+    - libgomp      # [linux]
+    - ninja
+  host:
+    - kokkos
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:26]: I guess an analogy might be:

- CUDA => Kokkos
- cuBLAS, cuSPARSE => Kokkos-kernels

There are other Kokkos extensions, like simd-math. As far as I know, they are all tightly coupled and for some the goal is to eventually merge them into the main Kokkos library.

File: recipes/kokkos-kernels/meta.yaml (hunk 2 above)
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:39]: I guess we could try building the examples and tests, would that work?

Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:39]: There is an error when compiling the tests, which I have reported [here](https://github.com/kokkos/kokkos-kernels/issues/1791).

**Review by leofang — APPROVED** [review]

File: recipes/kokkos-kernels/meta.yaml (hunk 2 above)
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:40]: Hi @leofang , what's the best way to get that file if the file is in `kokkos-kernels/example/batched_solve`. Should I copy from `$SRC_DIR/../work`?

Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml]: Found my answer to the previous questions, but my test fails because the c++ compiler is not found in the test phase. Where does one find g++ for a test?

File: recipes/kokkos-kernels/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  run:
+    - kokkos
+
+test:
+  source_files:
+    - example/batched_solve/team_GMRES.cpp
+  commands:
+    - gxx team_GMRES.cpp -o team_GMRES -I$PREFIX/include -L$PREFIX/lib -lkokkoscontainers -lkokkoscore -lkokkoskernels -fopenmp
```
Comment (leofang) [inline recipes/kokkos-kernels/meta.yaml:34]:
```suggestion
    - g++ team_GMRES.cpp -o team_GMRES -I$PREFIX/include -L$PREFIX/lib -lkokkoscontainers -lkokkoscore -lkokkoskernels -fopenmp
```

File: recipes/kokkos-kernels/meta.yaml (hunk 2 above)
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml]: Sorry for late reply, Vincent, have a long day... I think we just need to switch to `g++`.

File: recipes/kokkos-kernels/meta.yaml (hunk 5 above)
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml:34]: better:
```suggestion
    - $CXX team_GMRES.cpp -o team_GMRES -I$PREFIX/include -L$PREFIX/lib -lkokkoscontainers -lkokkoscore -lkokkoskernels -fopenmp
```

File: recipes/kokkos-kernels/meta.yaml (hunk 2 above)
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml]: For sure, no worries. I think I've tried that, but let me try again.

Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml]: Yes you did it in [153418a](https://github.com/conda-forge/staged-recipes/pull/22345/commits/153418a048e8bdfb997ae3bc67e81472a521d8fc), but you didn't add the compiler until [1a59fdc](https://github.com/conda-forge/staged-recipes/pull/22345/commits/1a59fdc5b276be5c9f9b2a3fd2bf73260704bb5d) 🙂

Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml]: linux-64 passed, great work Vincent!

Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml]: Thanks for your feedback ;)

File: recipes/kokkos-kernels/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  host:
+    - kokkos
+  run:
+    - kokkos
+
+test:
+  source_files:
+    - example/batched_solve
+  commands:
+    - $CXX example/batched_solve/team_GMRES.cpp -o team_GMRES -I$PREFIX/include -L$PREFIX/lib -lkokkoscontainers -lkokkoscore -lkokkoskernels -fopenmp
```
Comment (leofang) [inline recipes/kokkos-kernels/meta.yaml:34]: We may need to use C++17. It fails on macOS.
```suggestion
    - $CXX example/batched_solve/team_GMRES.cpp -o team_GMRES -std=c++17 -I$PREFIX/include -L$PREFIX/lib -lkokkoscontainers -lkokkoscore -lkokkoskernels -fopenmp
```

File: recipes/kokkos-kernels/meta.yaml (hunk 2 above)
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml]: All passed!

**Review by leofang — APPROVED** [review]

**Comment by leofang** [comment]
@BastianZim This is ready.

**Comment by BastianZim** [comment]
Thanks everyone!

File: recipes/kokkos-kernels/meta.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:16]: Hi Leo, I think I misunderstood you question here. I had in mind to match the versions of Kokkos and Kokkos-kernels, but I realize this has nothing to do with it. So I guess we might remove that, but I'm also wondering what's the best way to lock these together. Just 
```
  host:
    - kokkos =={{ version }}
```
?

[the remark this answers is not in the document]
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml:16]: @vincentmr I see. My question was "for those who depend on Kokkos Kernels, would they pin kokkos-kernels tightly that they use the same version at run time as at build time?" which is what this `run_exports` is about.

Your concern was that "Kokkos Kernels depends on Kokkos, will they be version-matched?" The answer is currently no. The [`run_exports` of Kokkos](https://github.com/conda-forge/kokkos-feedstock/blob/ba2a1102b1b9e99adc5afb1a62e4b93789822d81/recipe/meta.yaml#L19-L20) only guarantees that the run time and build time `kokkos` is the same version, but it's not necessary the same as the `kokkos-kernels` version. We should fix this in the feedstock.

[the remark this answers is not in the document]
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml:16]: In particular, note that kokkos v4.0.0 just came out. It seems `kokkos-kernels` v3.7.01 should be rebuilt, and the build number 0 should be marked as broken (since they depend on `kokkos` v4.0.0).

[the remark this answers is not in the document]
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml:16]: > It seems `kokkos-kernels` v3.7.01 should be rebuilt, 

Ah, I just saw that you already did it: https://github.com/conda-forge/kokkos-kernels-feedstock/pull/1, thanks!

> and the build number 0 should be marked as broken (since they depend on `kokkos` v4.0.0).

This needs to be done.

[the remark this answers is not in the document]
Comment (vincentmr) [reply recipes/kokkos-kernels/meta.yaml:16]: > In particular, note that kokkos v4.0.0 just came out a few weeks ago. It seems `kokkos-kernels` v3.7.01 should be rebuilt, and the build number 0 should be marked as broken (since they depend on `kokkos` v4.0.0).

I thought iterating the build number would forcefully deprecate lower build numbers. Ok, so let me try to do that. And thx for the feedback.

[the remark this answers is not in the document]
Comment (leofang) [reply recipes/kokkos-kernels/meta.yaml:16]: > I thought iterating the build number would forcefully deprecate lower build numbers.

Usually it's the case. But I've seen some complicated situations where the solver thinks it's better to pick the ones with a lower build number. Plus, in this case it's literally broken; we're not supposed to use Kokkos 4 to build Kokkos Kernels 3, unless there's strong compatibility guarantee from the Kokkos team.
